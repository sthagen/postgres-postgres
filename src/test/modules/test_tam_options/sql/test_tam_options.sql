CREATE EXTENSION test_tam_options;

-- encourage use of parallel plans
SET parallel_setup_cost = 0;
SET parallel_tuple_cost = 0;
SET min_parallel_table_scan_size = 0;
SET max_parallel_workers_per_gather = 4;

CREATE TABLE test (i int) USING heap_alter_options;

INSERT INTO test SELECT i FROM generate_series(1, 10000) i;
VACUUM ANALYZE test;

EXPLAIN (costs off)
SELECT * FROM test;

ALTER TABLE test SET (enable_parallel = OFF);

EXPLAIN (costs off)
SELECT * FROM test;

ALTER TABLE test SET (enable_parallel = ON);

EXPLAIN (costs off)
SELECT * FROM test;
