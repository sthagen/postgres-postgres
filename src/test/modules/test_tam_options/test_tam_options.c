/*--------------------------------------------------------------------------
 *
 * test_tam_options.c
 *		Test code for table access method reloptions.
 *
 * Copyright (c) 2024, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		src/test/modules/test_tam_options/test_tam_options.c
 *
 * -------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/reloptions.h"
#include "access/tableam.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(heap_alter_options_tam_handler);

/* An alternative relation options for heap */
typedef struct
{
	int32		vl_len_;		/* varlena header (do not touch directly!) */
	bool		enable_parallel;	/* enable parallel scans? */
} HeapAlterRdOptions;

static bytea *
heap_alter_reloptions(char relkind, Datum reloptions,
					  CommonRdOptions *common, bool validate)
{
	local_relopts relopts;
	HeapAlterRdOptions *result;

	Assert(relkind == RELKIND_RELATION ||
		   relkind == RELKIND_TOASTVALUE ||
		   relkind == RELKIND_MATVIEW);

	init_local_reloptions(&relopts, sizeof(HeapAlterRdOptions));
	add_local_bool_reloption(&relopts, "enable_parallel",
							 "enable parallel scan", true,
							 offsetof(HeapAlterRdOptions, enable_parallel));

	result = (HeapAlterRdOptions *) build_local_reloptions(&relopts,
														   reloptions,
														   validate);

	if (result != NULL && common != NULL)
	{
		common->parallel_workers = result->enable_parallel ? -1 : 0;
	}

	return (bytea *) result;
}

Datum
heap_alter_options_tam_handler(PG_FUNCTION_ARGS)
{
	static TableAmRoutine tam_routine;

	tam_routine = *GetHeapamTableAmRoutine();
	tam_routine.reloptions = heap_alter_reloptions;

	PG_RETURN_POINTER(&tam_routine);
}
