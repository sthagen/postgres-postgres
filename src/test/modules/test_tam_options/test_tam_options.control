comment = 'Test code for table access method reloptions'
default_version = '1.0'
module_pathname = '$libdir/test_tam_options'
relocatable = true
