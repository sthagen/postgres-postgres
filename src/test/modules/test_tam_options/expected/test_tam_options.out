CREATE EXTENSION test_tam_options;
-- encourage use of parallel plans
SET parallel_setup_cost = 0;
SET parallel_tuple_cost = 0;
SET min_parallel_table_scan_size = 0;
SET max_parallel_workers_per_gather = 4;
CREATE TABLE test (i int) USING heap_alter_options;
INSERT INTO test SELECT i FROM generate_series(1, 10000) i;
VACUUM ANALYZE test;
EXPLAIN (costs off)
SELECT * FROM test;
           QUERY PLAN            
---------------------------------
 Gather
   Workers Planned: 4
   ->  Parallel Seq Scan on test
(3 rows)

ALTER TABLE test SET (enable_parallel = OFF);
EXPLAIN (costs off)
SELECT * FROM test;
    QUERY PLAN    
------------------
 Seq Scan on test
(1 row)

ALTER TABLE test SET (enable_parallel = ON);
EXPLAIN (costs off)
SELECT * FROM test;
           QUERY PLAN            
---------------------------------
 Gather
   Workers Planned: 4
   ->  Parallel Seq Scan on test
(3 rows)

